### PR TITLE
Add EMP-SSL Loss

### PR DESCRIPTION
**Pull request by johnsutor**
This commit includes the code implementing the EMP-SSL Loss from [this repository](https://github.com/tsb0601/EMP-SSL/tree/main) and [this paper](https://arxiv.org/abs/2304.03977), and proposed in #1343 . The transform from the paper is not implemented, as it is simply a multi-crop transform based on the VICReg Transform (and a high-level implementation is given in the example code within the loss doc string)

### CODE REVIEW

**Comment by guarin** [comment]
Thanks for the PR! Will try to take a look ASAP.

File: lightly/loss/emp_ssl_loss.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        >>> # feed through encoder head
+        >>> z = torch.cat([model(s) for s in samples])
+        >>>
+        >>> # calculate loss
+        >>> loss = loss_fn(z)
+    """
+
+    def __init__(
+        self,
+        eps: float = 0.2,
```
Comment (guarin) [inline lightly/loss/emp_ssl_loss.py:81]: Let's call this `tcr_eps` to make it clear that it is a parameter for the `tcr_loss`.

File: lightly/loss/emp_ssl_loss.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        num_patches: int = 100,
+        tcr_coef: float = 1.0,
+        patch_coef: float = 200.0,
```
Comment (guarin) [inline lightly/loss/emp_ssl_loss.py:84]: Let's call this `inv_coef` to make it clear that is a parameter for the `invariance_loss`.

File: lightly/loss/emp_ssl_loss.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        >>>
+        >>> # calculate loss
+        >>> loss = loss_fn(z)
+    """
+
+    def __init__(
+        self,
+        eps: float = 0.2,
+        num_patches: int = 100,
+        tcr_coef: float = 1.0,
```
Comment (guarin) [inline lightly/loss/emp_ssl_loss.py:83]: I think we can drop both of these parameters. `num_patches` can be dropped because we can infer the number of patches based on the tensor passed to the loss function.

`tcr_coef` can be dropped because the number of coefficients of the total loss should be 1 - number of loss terms. The total loss `loss = patch_coef * patch_loss + tcr_coef * tcr_loss` could also be written as `loss = patch_coef / tcr_coef * patch_loss + tcr_loss`, thus "capturing" the value of the `tcr_coef` in the coefficient for the patch loss. Note that the paper uses only a single coefficient as well (lambda for the patch loss):
Screen Shot 2024-01-05 at 12 15 42

File: lightly/loss/emp_ssl_loss.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        self.patch_coef = patch_coef
+
+    def forward(self, z: Tensor) -> Tensor:
+        z_list = z.chunk(self.num_patches, dim=0)
+        contractive_loss = self.similarity_loss(z_list)
+        tcr_loss = torch.tensor(0.0, device=z.device, requires_grad=True)
+        for i in range(self.num_patches):
+            tcr_loss = tcr_loss + self.tcr(z_list[i])
+        tcr_loss = tcr_loss / self.num_patches
+        return Tensor(self.patch_coef * contractive_loss + self.tcr_coef * tcr_loss)
```
Comment (guarin) [inline lightly/loss/emp_ssl_loss.py:111]: If there are many views, we usually pass a list of views to the loss (see [DINOLoss](https://github.com/lightly-ai/lightly/blob/534cdf42a0feebe10a917bbdaacc6a533c1104fa/lightly/loss/dino_loss.py#L80-L85) or [VICRegLLoss](https://github.com/lightly-ai/lightly/blob/534cdf42a0feebe10a917bbdaacc6a533c1104fa/lightly/loss/vicregl_loss.py#L98-L104)). I suggest to change the function definition to `def forward(self, z_views: Sequence[Tensor]) -> Tensor`.

And if we go for the approach where the sub-losses are defined as functions the forward method can be rewritten as:
```
    def forward(self, z_views: Sequence[Tensor]) -> Tensor:
        # z has shape (num_views, batch_size, dim)
        z = torch.stack(z_views)
        return tcr_loss(z, eps=self.tcr_eps) + self.inv_coef * invariance_loss(z)
```

**Review by guarin — CHANGES_REQUESTED** [review]
Thanks a lot for the PR! I left some comments on how we could structure the code a bit to be more composable. Let me know what you think!

File: lightly/loss/emp_ssl_loss.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        z_stacked = torch.stack(z_list, dim=0)
+        z_avg = z_stacked.mean(dim=0)
+
+        z_sim = torch.tensor(0.0, device=z_stacked.device, requires_grad=True)
+        for i in range(num_patch):
+            z_sim = z_sim + F.cosine_similarity(z_stacked[i], z_avg, dim=1).mean()
+
+        z_sim = z_sim / num_patch
+
+        return -z_sim
```
[the remark this answers is not in the document]
Comment (johnsutor) [reply lightly/loss/emp_ssl_loss.py:38]: That seems fine by me, for simplicity, and is consistent with other loss functions like WMSE.

File: lightly/loss/emp_ssl_loss.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+class TotalCodingRate(nn.Module):
+    def __init__(self, eps: float = 0.2) -> None:
+        super(TotalCodingRate, self).__init__()
+        self.eps = eps
+
+    def compute_discrimn_loss(self, W: Tensor) -> Tensor:
+        """Discriminative Loss."""
+        d, b = W.shape
+        I = torch.eye(d, device=W.device)
+        scalar = d / (b * self.eps)
```
[the remark this answers is not in the document]
Comment (johnsutor) [reply lightly/loss/emp_ssl_loss.py:50]: I'll mention that, and I think it's fine to keep it as just epsilon instead of performing a redundant multiplication each time the loss is calculated.

File: lightly/loss/emp_ssl_loss.py (hunk 3 above)
Comment (johnsutor) [reply lightly/loss/emp_ssl_loss.py:83]: Sounds good. I was trying to stay faithful to the code implementation, but since the lambda term can be refactored as a combination of the coefficients, We can just have a lambda term.

**Comment by guarin** [comment]
@johnsutor would you like to make the changes or do you prefer if we make the changes on our side?

**Comment by johnsutor** [comment]
@guarin  I’m currently on vacation until Friday, I will push the updates then!

**Review by guarin — APPROVED** [review]
Awesome, thanks a lot for the changes! This is ready to be merged.
